tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs

```diff
@@ -0,0 +1,342 @@
+using Datadog.Trace.Debugger.Sink;
+using Datadog.Trace.Debugger.Snapshots;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Processors;
+using Datadog.Trace.RemoteConfigurationManagement;
+using Datadog.Trace.Telemetry;
+using Datadog.Trace.Telemetry.Metrics;
+using Datadog.Trace.Vendors.dnlib.Threading;
+using OperationCanceledException = System.OperationCanceledException;
+
+#nullable enable
+
+namespace Datadog.Trace.Debugger
+{
+    internal class DebuggerManager
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DebuggerManager));
+        private static readonly Lazy<DebuggerManager> _lazyInstance =
+            new(
+                () => new DebuggerManager(DebuggerSettings.FromDefaultSource(), ExceptionReplaySettings.FromDefaultSource()),
+                LazyThreadSafetyMode.ExecutionAndPublication);
+
+        private readonly SemaphoreSlim _semaphore;
+        private readonly CancellationTokenSource _cancellationToken;
+        private volatile bool _isShuttingDown;
+        private int _initialized;
+
+        private DebuggerManager(DebuggerSettings debuggerSettings, ExceptionReplaySettings exceptionReplaySettings)
+        {
+            _initialized = 0;
+            _isShuttingDown = false;
+            _semaphore = new SemaphoreSlim(1, 1);
+            DebuggerSettings = debuggerSettings;
+            ExceptionReplaySettings = exceptionReplaySettings;
+            _cancellationToken = new CancellationTokenSource();
+
+            var tracerManager = TracerManager.Instance;
+            try
+            {
+                DynamicInstrumentationHelper.ServiceName = TraceUtil.NormalizeTag(tracerManager.Settings.ServiceName ?? tracerManager.DefaultServiceName);
+            }
+            catch (Exception e)
+            {
+                DynamicInstrumentationHelper.ServiceName = tracerManager.DefaultServiceName;
+                Log.Error(e, "Could not set `DynamicInstrumentationHelper.ServiceName`.");
+            }
+        }
+
+        internal static DebuggerManager Instance => _lazyInstance.Value;
+
+        internal DebuggerSettings DebuggerSettings { get; private set; }
+
+        internal ExceptionReplaySettings ExceptionReplaySettings { get; }
+
+        internal DynamicInstrumentation? DynamicInstrumentation { get; private set; }
+
+        internal SpanCodeOrigin.SpanCodeOrigin? CodeOrigin { get; private set; }
+
+        internal IDebuggerUploader? SymbolsUploader { get; private set; }
+
+        internal ExceptionDebugging? ExceptionReplay { get; private set; }
+
+        internal string ServiceName => DynamicInstrumentationHelper.ServiceName;
+
+        // /!\ This method is called by reflection in the Samples.SampleHelpers
+        // If you remove it then you need to provide an alternative way to wait for the discovery service
+        private static async Task<bool> WaitForDiscoveryService(IDiscoveryService discoveryService)
```

Comment:
Should this method stay in a more "shared" location so other products can use it as well someday?